File: tests/tooling/perf/trees.fs
// The vanilla-F# twin of trees.fpp.
// The binary-trees allocation stress: build a deep tree, walk it, drop it,
// repeat. Almost every allocation dies young, which is the case a generational
// collector is built for and the case a bump-and-free C arena wins outright.
//
// Where avl measures a live structure being rebuilt, this measures the
// allocator and the collector on short-lived garbage.
module Trees

type Tree =
    | Leaf
    | Node of Tree * Tree

let rec make (d : int) : Tree =
    if d = 0 then Leaf
    else Node (make (d - 1), make (d - 1))

let rec check (t : Tree) : int =
    match t with
    | Leaf -> 1
    | Node (l, r) -> 1 + check l + check r

let maxDepth = 18
let minDepth = 4

[<EntryPoint>]
let main _ =
    let mutable acc = 0L
    // one long-lived tree, so the collector always has live data to trace
    let longLived = make maxDepth
    let mutable d = minDepth
    while d <= maxDepth do
        let iters = 1 <<< (maxDepth - d + minDepth)
        let mutable i = 0
        let mutable sum = 0
        while i < iters do
            sum <- sum + check (make d)
            i <- i + 1
        acc <- acc + int64 sum
        d <- d + 2
    acc <- acc + int64 (check longLived)
    printfn "%.0f" (float acc)
    0

File: tests/tooling/perf/trees.go
// The Go twin of trees.fpp: the same shape, run under the SAME wasmtime via
// GOOS=wasip1. It exists so the collector column is not read against C's
// arena reset — C here bump-allocates and drops a pointer, which no
// collected language can match. Go's is a mature generational-ish
// concurrent collector; this is what "a real GC on this workload" costs.
package main

import "fmt"

type node struct{ l, r *node }

func build(d int) *node {
	if d == 0 {
		return nil
	}
	return &node{build(d - 1), build(d - 1)}
}

func check(t *node) int {
	if t == nil {
		return 1
	}
	return 1 + check(t.l) + check(t.r)
}

const maxDepth = 18
const minDepth = 4

func main() {
	var acc int64
	longLived := build(maxDepth)
	for d := minDepth; d <= maxDepth; d += 2 {
		iters := 1 << (maxDepth - d + minDepth)
		sum := 0
		for i := 0; i < iters; i++ {
			sum += check(build(d))
		}
		acc += int64(sum)
	}
	acc += int64(check(longLived))
	fmt.Printf("%.0f\n", float64(acc))
}
